<a href="https://colab.research.google.com/github/sabrynbley/GB885_DiscussionCodingChallengeRepositories_Assignment_5/blob/main/GB885_Assignment_5_Bley_S.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Assignment 5 - Chef's Choice
### Sabryn Bley
### GENBUS 885 080

### **Configuration:**

Import the necessary libraries to load, inspect, and clean the data.

In [ ]:
import kagglehub      # needed to load dataset
import os             # needed to load dataset
import pandas as pd   # needed to investigate and clean dataset

### **Problem Definition:**


Fitness tracking data collected from applications such as Strava often contains missing values, inconsistent data formats, duplicate records, and invalid entries that can affect the accuracy of any analysis. Before meaningful insights can be generated, the dataset must be cleaned and validated to ensure it is reliable and consistent.

The goal of this project is to explore the quality of the running activity dataset and identify common data issues, such as missing values, duplicate records, inconsistent data types, formatting errors, and potential outliers. The dataset will then be cleaned by correcting or removing these issues to create a high-quality dataset that is suitable for future exploratory analysis and modeling.

By completing the data cleaning process, the resulting dataset will provide a reliable foundation for analyzing running performance, training habits, and other fitness-related trends in later stages of the project.

### **Dataset Design & Creation:**

The Strava Running Activity Data dataset was published on Kaggle by Ajit Jadhav and contains personal running activity data collected between March 2022 and December 2023. The data was exported using the Strava API, which records information from GPS-enabled fitness activities.

This dataset includes information about individual running activities, such as the activity date, distance, elapsed time, moving time, average pace or speed, elevation gain, heart rate (when available), and other workout metrics.

In [ ]:
# Download latest path to the data in Kaggle
path = kagglehub.dataset_download("ajitjadhav1/strava-running-activity-data")

# Get the data file name
file_name = os.listdir(path)[0]

# Define the full path to the data Excel
file_path = os.path.join(path, file_name)

# Load the data into a dataframe
running_df = pd.read_excel(file_path)

Using Colab cache for faster access to the 'strava-running-activity-data' dataset.


### **Data Inspection:**


In [ ]:
# Print the first 5 rows to visually check the data
running_df.head()

,Sr. no.,start_date_local,type,distance,moving_time,elapsed_time,total_elevation_gain,start_latlng,end_latlng,sport_type,start_date,timezone,achievement_count,kudos_count,comment_count,athlete_count,photo_count,average_speed,max_speed
0,1,2023-12-09T09:09:19Z,Run,10879.7,4023,4617,91.4,"[40.70327935740352, -73.99619171395898]","[40.69340907968581, -73.97922154515982]",Run,2023-12-09T14:09:19Z,(GMT-05:00) America/New_York,4,4,0,3,0,2.704,6.228
1,2,2023-12-07T17:31:50Z,Run,1304.4,722,62993,0.0,"[40.7220459356904, -74.03641730546951]","[40.719722136855125, -74.03257705271244]",Run,2023-12-07T22:31:50Z,(GMT-05:00) America/New_York,0,0,0,1,0,1.807,4.044
2,3,2023-12-03T09:18:13Z,Run,17503.0,7370,7462,68.4,"[40.74991073459387, -73.98751585744321]","[40.73538766242564, -73.9793517999351]",Run,2023-12-03T14:18:13Z,(GMT-05:00) America/New_York,2,5,0,4,0,2.375,6.778
3,4,2023-12-02T09:41:14Z,Run,3457.8,1791,2170,3.9,"[40.72203076444566, -74.03641068376601]","[40.71398631669581, -74.03902692720294]",Run,2023-12-02T14:41:14Z,(GMT-05:00) America/New_York,0,2,0,3,0,1.931,4.522
4,5,2023-12-01T17:06:05Z,Run,10108.2,4128,4221,6.3,"[40.72190687991679, -74.03624514117837]","[40.71669920347631, -74.03264125809073]",Run,2023-12-01T22:06:05Z,(GMT-05:00) America/New_York,4,5,0,1,0,2.449,5.462


In [ ]:
# View the dimensions of the dataset
print(f"Rows: {running_df.shape[0]}")
print(f"Columns: {running_df.shape[1]}")

Rows: 105
Columns: 19


In [ ]:
# Check the basic info (data types, number null cells, column names, ext.)
running_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 105 entries, 0 to 104
Data columns (total 19 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Sr. no.               105 non-null    int64  
 1   start_date_local      105 non-null    object 
 2   type                  105 non-null    object 
 3   distance              105 non-null    float64
 4   moving_time           105 non-null    int64  
 5   elapsed_time          105 non-null    int64  
 6   total_elevation_gain  105 non-null    float64
 7   start_latlng          105 non-null    object 
 8   end_latlng            105 non-null    object 
 9   sport_type            105 non-null    object 
 10  start_date            105 non-null    object 
 11  timezone              105 non-null    object 
 12  achievement_count     105 non-null    int64  
 13  kudos_count           105 non-null    int64  
 14  comment_count         105 non-null    int64  
 15  athlete_count         1

In [ ]:
# Check statistics of the data
running_df.describe()

,Sr. no.,distance,moving_time,elapsed_time,total_elevation_gain,achievement_count,kudos_count,comment_count,athlete_count,photo_count,average_speed,max_speed
count,105.000000,105.000000,105.000000,105.000000,105.000000,105.000000,105.000000,105.000000,105.000000,105.0,105.000000,105.000000
mean,53.000000,7067.378095,3554.019048,4753.238095,31.277143,3.076190,0.866667,0.028571,1.628571,0.0,2.197362,4.988343
std,30.454885,5785.399832,2627.200905,6956.141991,50.915702,3.657698,1.092515,0.167398,3.000824,0.0,0.797553,2.857700
min,1.000000,0.000000,85.000000,217.000000,0.000000,0.000000,0.000000,0.000000,1.000000,0.0,0.000000,0.000000
25%,27.000000,3181.300000,1880.000000,2042.000000,0.000000,0.000000,0.000000,0.000000,1.000000,0.0,2.015000,3.800000
50%,53.000000,5266.200000,2564.000000,3058.000000,5.000000,2.000000,0.000000,0.000000,1.000000,0.0,2.375000,4.586000
75%,79.000000,10137.900000,4768.000000,5181.000000,55.000000,4.000000,2.000000,0.000000,1.000000,0.0,2.566000,5.714000
max,105.000000,32192.300000,14536.000000,62993.000000,290.400000,17.000000,5.000000,1.000000,27.000000,0.0,6.085000,18.281000


In [ ]:
# Count missing values in each column
running_df.isnull().sum()

,0
Sr. no.,0
start_date_local,0
type,0
distance,0
moving_time,0
elapsed_time,0
total_elevation_gain,0
start_latlng,0
end_latlng,0
sport_type,0


In [ ]:
# Number of completely duplicated rows
duplicates = running_df.duplicated().sum()

print(f"Duplicate rows: {duplicates}")

Duplicate rows: 0


In [ ]:
# Get all columns that are objects
categorical = running_df.select_dtypes(include='object').columns

# Print out the values used in the columns
for col in categorical:
    print(running_df[col].value_counts(dropna=False).head(20))
    print("\n", "----", "\n")


start_date_local
2023-12-09T09:09:19Z    1
2023-12-07T17:31:50Z    1
2023-12-03T09:18:13Z    1
2023-12-02T09:41:14Z    1
2023-12-01T17:06:05Z    1
2023-11-29T17:22:51Z    1
2023-11-26T09:16:39Z    1
2023-11-09T17:46:38Z    1
2023-11-07T17:31:51Z    1
2023-11-04T16:20:18Z    1
2023-11-02T18:25:15Z    1
2023-11-01T18:08:35Z    1
2023-10-31T18:13:17Z    1
2023-10-29T18:03:13Z    1
2023-10-28T08:55:44Z    1
2023-10-23T09:11:55Z    1
2023-10-21T11:02:30Z    1
2023-10-19T19:22:33Z    1
2023-10-15T14:17:29Z    1
2023-10-12T18:27:56Z    1
Name: count, dtype: int64

 ---- 

type
Run        94
Workout     7
Walk        3
Hike        1
Name: count, dtype: int64

 ---- 

start_latlng
[]                                          9
[40.70327935740352, -73.99619171395898]     1
[40.7220459356904, -74.03641730546951]      1
[40.72203076444566, -74.03641068376601]     1
[40.74991073459387, -73.98751585744321]     1
[40.72185751050711, -74.03604313731194]     1
[40.74989422224462, -73.98728493601084]    

In [ ]:
# Activities with negative distances
print("Number of activites with negative distance:", len(running_df[running_df['distance'] < 0]))

# Activities with negative elapsed time
print("Number of activites with negative elapsed time:", len(running_df[running_df['elapsed_time'] < 0]))

Number of activites with negative distance: 0
Number of activites with negative elapsed time: 0


### **Data Cleaning:**

**Fix Column Data Types:**


The date columns were imported as strings, so convert them to be datetime objects. This will allow for easier use in later analysis.

In [ ]:
# Convert to datetime objects
running_df['start_date'] = pd.to_datetime(running_df['start_date'])
running_df['start_date_local'] = pd.to_datetime(running_df['start_date_local'])

**Remove Unneeded Activites:**

Out of 105 records, only 11 of the activities are not runs. We will focus on the running activites, so we can remove the other activity records.

In [ ]:
# Remove walks, workouts, and hikes
running_df = running_df[~running_df['type'].isin(['Walk', 'Hike', 'Workout'])]

**Remove Missing Lat & Lon Pairs:**

There were initially 9 records missing (there are not NA values, but they are empty pairs) a start and end latitude and longitude coordinates. For consistency in quality, we will remove the records that are missing this.

In [ ]:
running_df = running_df[
    (running_df['start_latlng'] != '[]') &
    (running_df['end_latlng'] != '[]')
]

**Remove Unneeded Columns:**

The sports type and type columns are essentially duplicates of each other. We only need one of them. Sr. no. is just what number in the table the record is. It doesn't hold any contextual value. Additionally, there are metrics for the activity's post on Strava's platform. It doesn't provide value for analysis of the activity itself.

In [ ]:
running_df = running_df.drop(columns=['sport_type', 'Sr. no.', 'achievement_count', 'kudos_count', 'comment_count', 'athlete_count', 'photo_count'])

**Rename Columns:**

The Kaggle dataset page (https://www.kaggle.com/datasets/ajitjadhav1/strava-running-activity-data) has some information about the units the values are measured in. To better keep track of that information, we can include it in the column names.

In [ ]:
running_df.rename(columns={'distance': 'distance_in_meters',
                           'moving_time': 'moving_time_in_secs',
                           'elapsed_time': 'elapsed_time_in_secs',
                           'total_elevation_gain': 'total_elevation_gain_in_meters',
                           'start_date': 'start_date_utc',
                           'average_speed': 'average_speed_in_mps',
                           'max_speed': 'max_speed_in_mps'}, inplace=True)

### **Review the Final Data:**

In [ ]:
# Print the first 5 rows to visually check the data
running_df.head()

,start_date_local,type,distance_in_meters,moving_time_in_secs,elapsed_time_in_secs,total_elevation_gain_in_meters,start_latlng,end_latlng,start_date_utc,timezone,average_speed_in_mps,max_speed_in_mps
0,2023-12-09 09:09:19+00:00,Run,10879.7,4023,4617,91.4,"[40.70327935740352, -73.99619171395898]","[40.69340907968581, -73.97922154515982]",2023-12-09 14:09:19+00:00,(GMT-05:00) America/New_York,2.704,6.228
1,2023-12-07 17:31:50+00:00,Run,1304.4,722,62993,0.0,"[40.7220459356904, -74.03641730546951]","[40.719722136855125, -74.03257705271244]",2023-12-07 22:31:50+00:00,(GMT-05:00) America/New_York,1.807,4.044
2,2023-12-03 09:18:13+00:00,Run,17503.0,7370,7462,68.4,"[40.74991073459387, -73.98751585744321]","[40.73538766242564, -73.9793517999351]",2023-12-03 14:18:13+00:00,(GMT-05:00) America/New_York,2.375,6.778
3,2023-12-02 09:41:14+00:00,Run,3457.8,1791,2170,3.9,"[40.72203076444566, -74.03641068376601]","[40.71398631669581, -74.03902692720294]",2023-12-02 14:41:14+00:00,(GMT-05:00) America/New_York,1.931,4.522
4,2023-12-01 17:06:05+00:00,Run,10108.2,4128,4221,6.3,"[40.72190687991679, -74.03624514117837]","[40.71669920347631, -74.03264125809073]",2023-12-01 22:06:05+00:00,(GMT-05:00) America/New_York,2.449,5.462
